In [1]:
# Import the modules
import numpy as np
import pandas as pd
from pathlib import Path
from sklearn.metrics import balanced_accuracy_score, confusion_matrix, classification_report

## Split the Data into Training and Testing Sets

### Step 1: Read the `lending_data.csv` data from the `Resources` folder into a Pandas DataFrame.

In [2]:
# Read the CSV file from the Resources folder into a Pandas DataFrame
lending_data_df = pd.read_csv("Resources/lending_data.csv")

# Review the DataFrame
lending_data_df.head()

loan_size  interest_rate  borrower_income  debt_to_income  num_of_accounts  \
0    10700.0          7.672            52800        0.431818                5   
1     8400.0          6.692            43600        0.311927                3   
2     9000.0          6.963            46100        0.349241                3   
3    10700.0          7.664            52700        0.430740                5   
4    10800.0          7.698            53000        0.433962                5   

   derogatory_marks  total_debt  loan_status  
0                 1       22800            0  
1                 0       13600            0  
2                 0       16100            0  
3                 1       22700            0  
4                 1       23000            0

### Step 2: Create the labels set (`y`)  from the “loan_status” column, and then create the features (`X`) DataFrame from the remaining columns.

In [3]:
# Separate the data into labels and features

# Separate the y variable, the labels
y = lending_data_df['loan_status']

# Separate the X variable, the features
X = lending_data_df.drop(columns=['loan_status'])

In [4]:
# Review the y variable Series
y[:5]

0    0
1    0
2    0
3    0
4    0
Name: loan_status, dtype: int64

In [5]:
# Review the X variable DataFrame
X.head()

loan_size  interest_rate  borrower_income  debt_to_income  num_of_accounts  \
0    10700.0          7.672            52800        0.431818                5   
1     8400.0          6.692            43600        0.311927                3   
2     9000.0          6.963            46100        0.349241                3   
3    10700.0          7.664            52700        0.430740                5   
4    10800.0          7.698            53000        0.433962                5   

   derogatory_marks  total_debt  
0                 1       22800  
1                 0       13600  
2                 0       16100  
3                 1       22700  
4                 1       23000

### Step 3: Check the balance of the labels variable (`y`) by using the `value_counts` function.

In [6]:
# Check the balance of our target values
y.value_counts()

0    75036
1     2500
Name: loan_status, dtype: int64

### Step 4: Split the data into training and testing datasets by using `train_test_split`.

In [7]:
# Import the train_test_learn module
from sklearn.model_selection import train_test_split

# Split the data using train_test_split
# Assign a random_state of 1 to the function
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=1)

## Create a Logistic Regression Model with the Original Data

###  Step 1: Fit a logistic regression model by using the training data (`X_train` and `y_train`).

In [8]:
# Import the LogisticRegression module from SKLearn
from sklearn.linear_model import LogisticRegression

# Instantiate the Logistic Regression model
# Assign a random_state parameter of 1 to the model
logistic_regression_model =LogisticRegression(random_state = 1)

# Fit the model using training data
lr_model= logistic_regression_model.fit(X_train, y_train)

### Step 2: Save the predictions on the testing data labels by using the testing feature data (`X_test`) and the fitted model.

In [10]:
# Make a prediction using the testing data
testing_predictions = lr_model.predict(X_test)

### Step 3: Evaluate the model’s performance by doing the following:

* Calculate the accuracy score of the model.

* Generate a confusion matrix.

* Print the classification report.

In [12]:
# Print the balanced_accuracy score of the model
print(f"Training Data Score: {lr_model.score(X_train, y_train)}")
print(f"Testing Data Score: {lr_model.score(X_test, y_test)}")


# Display the accuracy score for the test dataset.
print(f"Balanced Accuracy Score: {balanced_accuracy_score(y_test,testing_predictions)}")

Training Data Score: 0.9921240885954051
Testing Data Score: 0.9918489475856377
Balanced Accuracy Score: 0.9520479254722232


In [14]:
results = pd.DataFrame({"Prediction": testing_predictions, "Actual": y_test}).reset_index(drop=True)
results.head(10)

Prediction  Actual
0           0       0
1           0       0
2           0       0
3           0       0
4           0       0
5           0       0
6           0       0
7           0       0
8           0       0
9           0       0

In [16]:
# Generate a confusion matrix for the model

#Import the confusion_matrix module from SKLearn
from sklearn.metrics import confusion_matrix

testing_matrix = confusion_matrix(y_test, testing_predictions)
print(testing_matrix)

[[18663   102]
 [   56   563]]


In [17]:
# Print the classification report for the model
testing_report = classification_report(y_test, testing_predictions)

print(testing_report)

              precision    recall  f1-score   support

           0       1.00      0.99      1.00     18765
           1       0.85      0.91      0.88       619

    accuracy                           0.99     19384
   macro avg       0.92      0.95      0.94     19384
weighted avg       0.99      0.99      0.99     19384



### Step 4: Answer the following question.

**Question:** How well does the logistic regression model predict both the `0` (healthy loan) and `1` (high-risk loan) labels?

**Answer:** Based on the classification report, the logistic regression model seems to predict both the 0 (healthy loan) and 1 (high-risk loan) labels quite well. The precision of the model is high for both classes, with a value of 1.00 for class 0 and 0.85 for class 1. This indicates that when the model predicts a loan as healthy (class 0), it is correct 100% of the time, and when it predicts a loan as high-risk (class 1), it is correct 85% of the time.

The recall of the model is also high for both classes, with a value of 0.99 for class 0 and 0.91 for class 1. This indicates that the model is able to correctly identify a large proportion of the loans that are actually healthy (class 0) and high-risk (class 1).

The f1-score, which is a weighted average of precision and recall, is high for both classes, with a value of 1.00 for class 0 and 0.88 for class 1. This indicates that the model is able to achieve a good balance between precision and recall for both classes.

However, it's important to note that the dataset is imbalanced, meaning that there are more healthy loans than high-risk loans. This could potentially affect the model's performance, and it's essential to look at resampled data to assess the model's performance accurately.

In terms of the cost of misclassification, both precision and recall are important metrics to consider. A high precision score is crucial to minimize false positives, which could result in a loss of potential customers. On the other hand, a high recall score is essential to minimize false negatives, which could lead to significant financial losses.

Overall, the model's accuracy of 0.99 and the macro and weighted averages suggest that it is performing well across both classes. However, it's essential to keep in mind the potential impact of an imbalanced dataset and to evaluate the model's performance using resampled data to get a more accurate assessment.


## Predict a Logistic Regression Model with Resampled Training Data

### Step 1: Use the `RandomOverSampler` module from the imbalanced-learn library to resample the data. Be sure to confirm that the labels have an equal number of data points. 

In [18]:
# Import the RandomOverSampler module form imbalanced-learn
from imblearn.over_sampling import RandomOverSampler

# Instantiate the random oversampler model
# # Assign a random_state parameter of 1 to the model
ROS = RandomOverSampler(random_state=1)

# Fit the original training data to the random_oversampler model
X_resampled, y_resampled= ROS.fit_resample(X_train, y_train)

In [19]:
# Count the distinct values of the resampled labels data
y_resampled.value_counts()

0    56271
1    56271
Name: loan_status, dtype: int64

### Step 2: Use the `LogisticRegression` classifier and the resampled data to fit the model and make predictions.

In [20]:
# Instantiate the Logistic Regression model
# Assign a random_state parameter of 1 to the model
ros_model = LogisticRegression(solver='lbfgs', random_state=1)

# Fit the model using the resampled training data
ros_model.fit(X_resampled, y_resampled)

# Make a prediction using the testing data
y_pred = ros_model.predict(X_test)

### Step 3: Evaluate the model’s performance by doing the following:

* Calculate the accuracy score of the model.

* Generate a confusion matrix.

* Print the classification report.

In [21]:
# Print the balanced_accuracy score of the model 
print(f"Training Data Score: {ros_model.score(X_train, y_train)}")
print(f"Testing Data Score: {ros_model.score(X_test, y_test)}")
print(f"Balanced Accuracy Score: {balanced_accuracy_score(y_test, y_pred)}")

Training Data Score: 0.9945659650570917
Testing Data Score: 0.9938093272802311
Balanced Accuracy Score: 0.9936781215845847


In [22]:
# Generate a confusion matrix for the model
from sklearn.metrics import confusion_matrix
ros_testing_matrix = confusion_matrix(y_test, y_pred)
print(ros_testing_matrix)

[[18649   116]
 [    4   615]]


In [23]:
# Print the classification report for the model
ros_testing_report = classification_report(y_test, y_pred)

print(ros_testing_report)

              precision    recall  f1-score   support

           0       1.00      0.99      1.00     18765
           1       0.84      0.99      0.91       619

    accuracy                           0.99     19384
   macro avg       0.92      0.99      0.95     19384
weighted avg       0.99      0.99      0.99     19384



### Step 4: Answer the following question

**Question:** How well does the logistic regression model, fit with oversampled data, predict both the `0` (healthy loan) and `1` (high-risk loan) labels?

**Answer:** the logistic regression model fit with oversampled data outperforms the original model in terms of predicting high-risk loans. Specifically, the precision score for high-risk loans is higher in the oversampled model (0.84) compared to the original model (0.85). This means that the oversampled model is better at identifying which loans are actually high-risk.

Additionally, the recall score for high-risk loans is also higher in the oversampled model (0.99) compared to the original model (0.91). This means that the oversampled model is better at correctly identifying a large proportion of the actual high-risk loans.

Overall, the oversampled model seems to be more effective at predicting high-risk loans, which is important in reducing the risk of financial losses due to loan defaults.